orttraining/orttraining/core/optimizer/pythonop_rewriter.cc

```diff
@@ -0,0 +1,114 @@
+        int64_t tensor_index = 0;
+        const size_t all_input_count = input_convention.size();
+        position_to_tensor_index.reserve(all_input_count);
+        for (size_t i = 0; i < all_input_count; ++i) {
+          if (input_convention[i] == 'd') {
+            position_to_tensor_index[i] = tensor_index;
+            ++tensor_index;
+          }
+        }
+
+        for (size_t i = 1; i < fw_tensor_output_to_tensor_input_reuse_map.size(); ++i) {
+          if (fw_all_output_to_tensor_input_reuse_map[i - 1] != -1) {
+            ORT_ENFORCE(fw_all_output_to_tensor_input_reuse_map[i - 1] < static_cast<int64_t>(all_input_count),
+                        "PythonOp input alias function output index out of range. func_name: ", func_name, " ",
+                        fw_all_output_to_tensor_input_reuse_map[i - 1], " >= ", all_input_count);
+            fw_tensor_output_to_tensor_input_reuse_map[i] =
+                position_to_tensor_index.at(fw_all_output_to_tensor_input_reuse_map[i - 1]);
+          }
+        }
+
+        node.AddAttribute("tensor_reuse_map", fw_tensor_output_to_tensor_input_reuse_map);
+      }
+
+      {
+        // Handle backward input alias map.
+        auto& output_convention = input_convention;
+        ORT_ENFORCE(bw_all_output_to_tensor_input_reuse_map.size() == output_convention.size(),
+                    "PythonOpGrad input alias function output count mismatch. func_name: ", func_name, " ",
+                    bw_all_output_to_tensor_input_reuse_map.size(), " != ", output_convention.size());
+
+        std::vector<int64_t> bw_tensor_output_to_tensor_input_reuse_map =
+            std::vector<int64_t>(node.InputDefs().size(), -1);
+        size_t tensor_output_index = 0;
+        for (size_t i = 0; i < output_convention.size(); ++i) {
+          if (output_convention[i] == 'd') {
+            ORT_ENFORCE(tensor_output_index < bw_tensor_output_to_tensor_input_reuse_map.size(),
+                        "PythonOpGrad input alias function output count mismatch. func_name: ", func_name, " ",
+                        tensor_output_index, " >= ", bw_tensor_output_to_tensor_input_reuse_map.size());
+            // input index shift by 1 to skip the context
+            bw_tensor_output_to_tensor_input_reuse_map[tensor_output_index] =
+                bw_all_output_to_tensor_input_reuse_map[i] == -1 ? -1 : bw_all_output_to_tensor_input_reuse_map[i] + 1;
+            ++tensor_output_index;
+          }
+        }
+        node.AddAttribute("bw_tensor_reuse_map", bw_tensor_output_to_tensor_input_reuse_map);
+      }
+
+      modified = true;
+    }
+  }
+
+  if (modified)
+    rule_effect = RewriteRuleEffect::kUpdatedCurrentNode;
+#else
+  ORT_UNUSED_PARAMETER(node);
+  ORT_UNUSED_PARAMETER(rule_effect);
+
+#endif
+  return Status::OK();
+}
+
+bool PythonOpRewriter::SatisfyCondition(const Graph&, const Node&, const logging::Logger&) const {
+  return true;
```

Comment:
Why not move the optimizer itself behind ifdef? This way inference build for example will never even pick up this code. 
Another suggestion is, if you continue to keep the optimizer then use ifdef here and return true when torch interop is enabled and false when not.

Reply:
Used ifdef for the whole files.